File: spring-ai-alibaba-jmanus/src/main/java/com/alibaba/cloud/ai/example/manus/recorder/entity/PlanExecutionRecord.java
/*
 * Copyright 2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.alibaba.cloud.ai.example.manus.recorder.entity;

import com.alibaba.cloud.ai.example.manus.planning.model.vo.UserInputWaitState;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import com.fasterxml.jackson.datatype.jsr310.deser.LocalDateTimeDeserializer;
import com.fasterxml.jackson.datatype.jsr310.ser.LocalDateTimeSerializer;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Plan execution record class for tracking and recording detailed information about
 * PlanningFlow execution process.
 *
 * Data structure is divided into four main parts:
 *
 * 1. Basic Info - id: Unique identifier of the record - currentPlanId: Unique identifier
 * of the plan - title: Plan title - startTime: Execution start time - endTime: Execution
 * end time - userRequest: User's original request
 *
 * 2. Plan Structure - steps: Plan step list - currentStepIndex: Current step index being
 * executed
 *
 * 3. Execution Data - agentExecutionSequence: Records of execution for each smart agent
 *
 * 4. Execution Result - completed: Whether completed - summary: Execution summary
 */
public class PlanExecutionRecord {

	// Unique identifier for the record
	private Long id;

	// Unique identifier for the current plan
	private String currentPlanId;

	// Plan title
	private String title;

	// User's original request
	private String userRequest;

	// Timestamp when execution started
	@JsonSerialize(using = LocalDateTimeSerializer.class)
	@JsonDeserialize(using = LocalDateTimeDeserializer.class)
	private LocalDateTime startTime;

	// Timestamp when execution ended
	@JsonSerialize(using = LocalDateTimeSerializer.class)
	@JsonDeserialize(using = LocalDateTimeDeserializer.class)
	private LocalDateTime endTime;

	// List of plan steps
	private List<String> steps;

	// Current step index being executed
	private Integer currentStepIndex;

	// Whether completed
	private boolean completed;

	// Execution summary
	private String summary;

	// List to maintain the sequence of agent executions
	private List<AgentExecutionRecord> agentExecutionSequence;

	// Field to store user input wait state
	private UserInputWaitState userInputWaitState;

	// Actual calling model
	private String modelName;

	/**
	 * Default constructor for Jackson and other frameworks.
	 */
	public PlanExecutionRecord() {
		this.steps = new ArrayList<>();
		this.completed = false;
		this.agentExecutionSequence = new ArrayList<>();
		this.id = generateId();
	}

	/**
	 * Constructor for creating a new execution record
	 * @param currentPlanId The unique identifier for the current plan.
	 */
	public PlanExecutionRecord(String currentPlanId) {
		this.currentPlanId = currentPlanId;
		this.steps = new ArrayList<>();
		this.startTime = LocalDateTime.now();
		this.completed = false;
		this.agentExecutionSequence = new ArrayList<>();
		this.id = generateId();
	}

	/**
	 * Add an execution step
	 * @param step Step description
	 */
	public void addStep(String step) {
		this.steps.add(step);
	}

	/**
	 * Add agent execution record
	 * @param record Execution record
	 */
	public void addAgentExecutionRecord(AgentExecutionRecord record) {
		this.agentExecutionSequence.add(record);
	}

	/**
	 * Get agent execution records sorted by execution order
	 * @return List of execution records
	 */
	public List<AgentExecutionRecord> getAgentExecutionSequence() {
		return agentExecutionSequence;
	}

	public void setAgentExecutionSequence(List<AgentExecutionRecord> agentExecutionSequence) {
		this.agentExecutionSequence = agentExecutionSequence;
	}

	/**
	 * Complete execution and set end time
	 */
	public void complete(String summary) {
		this.endTime = LocalDateTime.now();
		this.completed = true;
		this.summary = summary;
	}

	/**
	 * Generate unique ID if not already set
	 * @return Generated or existing ID
	 */
	private Long generateId() {
		if (this.id == null) {
			// Use combination of timestamp and random number to generate ID
			long timestamp = System.currentTimeMillis();
			int random = (int) (Math.random() * 1000000);
			this.id = timestamp * 1000 + random;
		}
		return this.id;
	}

	// Getters and Setters

	public Long getId() {
		// Ensure ID is generated when accessing
		if (this.id == null) {
			this.id = generateId();
		}
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public String getCurrentPlanId() {
		return currentPlanId;
	}

	public void setCurrentPlanId(String currentPlanId) {
		this.currentPlanId = currentPlanId;
	}

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		this.title = title;
	}

	public String getUserRequest() {
		return userRequest;
	}

	public void setUserRequest(String userRequest) {
		this.userRequest = userRequest;
	}

	public LocalDateTime getStartTime() {
		return startTime;
	}

	public void setStartTime(LocalDateTime startTime) {
		this.startTime = startTime;
	}

	public UserInputWaitState getUserInputWaitState() {
		return userInputWaitState;
	}

	public void setUserInputWaitState(UserInputWaitState userInputWaitState) {
		this.userInputWaitState = userInputWaitState;
	}

	public LocalDateTime getEndTime() {
		return endTime;
	}

	public void setEndTime(LocalDateTime endTime) {
		this.endTime = endTime;
	}

	public List<String> getSteps() {
		return steps;
	}

	public void setSteps(List<String> steps) {
		this.steps = steps;
	}

	public Integer getCurrentStepIndex() {
		return currentStepIndex;
	}

	public void setCurrentStepIndex(Integer currentStepIndex) {
		this.currentStepIndex = currentStepIndex;
	}

	public boolean isCompleted() {
		return completed;
	}

	public void setCompleted(boolean completed) {
		this.completed = completed;
	}

	public String getSummary() {
		return summary;
	}

	public void setSummary(String summary) {
		this.summary = summary;
	}

	public String getModelName() {
		return modelName;
	}

	public void setModelName(String modelName) {
		this.modelName = modelName;
	}

}

File: spring-ai-alibaba-jmanus/ui-vue3/src/types/plan-execution-record.ts
/*
 * Copyright 2025 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * TypeScript interfaces matching Java backend data structures
 *
 * Hierarchy:
 * PlanExecutionRecord (Main)
 *   └── agentExecutionSequence: AgentExecutionRecord[]
 *       └── subPlanExecutionRecords: PlanExecutionRecord[] (recursive)
 */

/**
 * Execution status enum matching Java ExecutionStatus
 * Maps to Java: com.alibaba.cloud.ai.manus.recorder.entity.vo.ExecutionStatus
 */
export type ExecutionStatus = 'IDLE' | 'RUNNING' | 'FINISHED'

/**
 * Maps to Java: com.alibaba.cloud.ai.manus.recorder.entity.vo.ActToolInfo
 */
export interface ActToolInfo {
  /** Name of the tool */
  name?: string

  /** Description of the tool */
  parameters?: string

  /** Result of tool call */
  result?: string

  /** ID of the tool */
  id?: string
}

/**
 * Maps to Java: com.alibaba.cloud.ai.manus.recorder.entity.ThinkActRecord
 *
 * Records the thinking and action process of an agent in a single execution step.
 */
export interface ThinkActRecord {
  /** Unique identifier of the record */
  id?: number

  /** ID of parent execution record, linked to AgentExecutionRecord */
  parentExecutionId?: number

  /** Timestamp when thinking started */
  thinkStartTime?: string

  /** Timestamp when thinking completed */
  thinkEndTime?: string

  /** Timestamp when action started */
  actStartTime?: string

  /** Timestamp when action completed */
  actEndTime?: string

  /** Input context for the thinking process */
  thinkInput?: string

  /** Output result of the thinking process */
  thinkOutput?: string

  /** Whether thinking determined that action is needed */
  actionNeeded?: boolean

  /** Description of the action to be taken */
  actionDescription?: string

  /** Result of action execution */
  actionResult?: string

  /** Status of this think-act cycle */
  status?: ExecutionStatus

  /** Error message if the cycle encountered problems */
  errorMessage?: string

  /** Tool name used for action (if applicable) */
  toolName?: string

  /** Tool parameters used for action (serialized, if applicable) */
  toolParameters?: string

  /** Tool call information */
  actToolInfoList?: ActToolInfo[]

  /** Sub-plan execution record for tool calls that create new execution plans */
  subPlanExecutionRecord?: PlanExecutionRecord
}

/**
 * Maps to Java: com.alibaba.cloud.ai.manus.recorder.entity.vo.AgentExecutionRecord
 *
 * Unified agent execution record class for tracking and recording detailed information about
 * BaseAgent execution process. This class combines the functionality of both Simple and Detail
 * versions to provide a single, comprehensive record structure.
 */
export interface AgentExecutionRecord {
  /** Unique identifier of the record */
  id?: number

  /** Step ID this record belongs to */
  stepId?: string

  /** Conversation ID this record belongs to */
  conversationId?: string

  /** Name of the agent that created this record */
  agentName?: string

  /** Description information of the agent */
  agentDescription?: string

  /** Timestamp when execution started */
  startTime?: string

  /** Timestamp when execution ended */
  endTime?: string

  /** Maximum allowed number of steps */
  maxSteps?: number

  /** Current execution step number */
  currentStep?: number

  /** Execution status (IDLE, RUNNING, FINISHED) */
  status?: ExecutionStatus

  /** Request content for agent execution */
  agentRequest?: string

  /** Execution result */
  result?: string

  /** Error message if execution encounters problems */
  errorMessage?: string

  /** Model name called */
  modelName?: string

  /** List of ThinkActRecord for detailed execution process */
  thinkActSteps?: ThinkActRecord[]

  /** Sub-plan execution records for this agent */
  subPlanExecutionRecords?: PlanExecutionRecord[]
}

/**
 * Maps to Java: com.alibaba.cloud.ai.manus.runtime.entity.vo.UserInputWaitState
 *
 * User input wait state for handling user interaction during plan execution.
 */
export interface UserInputWaitState {
  /** Plan ID */
  planId?: string

  /** Title to display to user */
  title?: string

  /** Whether the plan is currently waiting for user input */
  waiting?: boolean

  /** Form description */
  formDescription?: string

  /** Form input fields as map list (matching Java structure) */
  formInputs?: Array<Record<string, string>>
}

/**
 * Maps to Java: com.alibaba.cloud.ai.manus.recorder.entity.vo.PlanExecutionRecord
 *
 * Plan execution record class for tracking and recording detailed information about
 * PlanningFlow execution process.
 *
 * This is the main structure returned by /api/executor/details/{planId}
 */
export interface PlanExecutionRecord {
  /** Unique identifier for the record */
  id?: number

  /**
   * Unique identifier for the current plan
   * Maps to Java field: currentPlanId
   * Frontend uses this as: planId
   */
  currentPlanId: string

  /** Root plan ID for sub-plans (null for main plans) */
  rootPlanId?: string

  /** Parent plan ID for sub-plans (null for root plans) */
  parentPlanId?: string

  /** Tool call ID that triggered this plan (for sub-plans) */
  toolCallId?: string

  /** Plan title */
  title?: string

  /** User's original request */
  userRequest?: string

  /** Timestamp when execution started */
  startTime?: string

  /** Timestamp when execution ended */
  endTime?: string



  /** Current step index being executed */
  currentStepIndex?: number

  /** Whether completed */
  completed?: boolean

  /** Execution summary */
  summary?: string

  /**
   * List to maintain the sequence of agent executions
   * This is the main data source for step-by-step execution details
   * Uses AgentExecutionRecord (unified version)
   */
  agentExecutionSequence?: AgentExecutionRecord[]

  /** Field to store user input wait state */
  userInputWaitState?: UserInputWaitState

  /** Actual calling model */
  modelName?: string

  /** Parent tool call information that triggered this sub-plan (for sub-plan detail displaying) */
  parentActToolCall?: ActToolInfo

  // Additional computed fields for frontend compatibility
  /** Current execution status */
  status?: 'pending' | 'running' | 'completed' | 'failed' | 'paused'

  /** Execution progress percentage (0-100) */
  progress?: number

  /** Current progress text description */
  progressText?: string

  /** Detailed execution result */
  result?: string

  /** Error message if execution failed */
  error?: string

  /** Additional message */
  message?: string

  /** Creation timestamp */
  createdAt?: string

  /** Last update timestamp */
  updatedAt?: string

  /** Additional metadata */
  metadata?: Record<string, any>
}

// API response type for getDetails method
export type PlanExecutionRecordResponse = PlanExecutionRecord | null


